### ISSUE TITLE
reinforcement learning dataloading and algorithms

### ISSUE THREAD

**Issue by alexge233** [pytorch/pytorch#34223]
## 🚀 Feature
Implement a dataloading functionality for reinforcement learning state, action pairs, with assigned policy scores, transitional probabilities and rewards.
Implement a set of gradient algorithms (qlearning, actor-critic, etc) in order to populate policy maps and use them or export them for other models.

## Motivation

I feel there is a gap in pytorch, which doesn't address a relatively interesting field of machine learning. I understand that reinforcement learning falls outside deep learning, and that simple examples of DQN already exist in tutorials, blogs and such, but I feel that impementing a GPU and CPU based dataloader which is easy to use and understand, combined with the algorithms that calculate the policy values and maps, would benefit pytorch as well as those who already use pytorch for a variety of reasons. I think that there's a lot of software domains where this would be applicable.

## Pitch

1. develop a dataloader class or set of classes which deal with deterministic state-action spaces and with probabilistic state-action spaces.
2. develop a set of reinforcement learning algorithm classes which can be applied to those dataloaders, and populate policy values, based on rewards and such

## Alternatives

No much really.

## Additional context

I think that pytorch has grown to be a wonderful tool used in so many different places, and that reinforcement learning seems to be missing. I know theres a tutorial for DQN, and I know that a lot of the algorithms are trivial and simple, yet it seems to me that users rely on pytorch as a simple and easy to use tool for machine learning, and as such don't always have the time (or know-how) on how to implement that functionality.

**Just to be clear** I propose **I implement this** with some guidance, as I've already done this in C, C++ and Python before. I've got a few ideas on how to do this with the underlying C++ aten and autograd libraries and then move on to the python bindings.

cc @SsnL

**Comment by alexge233** [comment]
Just a bump, in order to get more people's opinions.

**Comment by soumith** [comment]
@alexge233 I reviewed this proposal out of interest today.

I am still unclear on your plan, on two fronts:

1. Are you proposing a new DataLoader class that is specialized for state-action spaces? If so, why is the existing one insufficient? From what I can tell, it has the concept of a Dataset, and a Sampler or you can use an IterableDataset that gives more flexibility

2. Do you propose to have this in the core of pytorch/pytorch ? If yes, to me this is not obvious as to why.

If it's a matter of discovering high-quality RL repos that have standard common data loading implemented, I wonder if we should put together some pointers to well-known RL packages in the PyTorch.

For example:
- https://rlpyt.readthedocs.io/en/latest/
- https://github.com/facebookresearch/torchbeast
- https://github.com/p-christ/Deep-Reinforcement-Learning-Algorithms-with-PyTorch
- https://github.com/jingweiz/pytorch-rl
- https://github.com/ikostrikov/pytorch-a3c and https://github.com/denisyarats/drq

are a few high-quality ones I know.

**Comment by SsnL** [comment]
Yeah I generally agree with @soumith . It's unclear what your request mean 

> develop a dataloader class or set of classes which deal with deterministic state-action spaces and with probabilistic state-action spaces.

As @soumith said, the provided APIs are general enough that I don't see why state and/or action space, deterministic or not, can be a problem. The only thing I can think of that requires care is generating on-policy data with CUDA policy, where `spawn` must be used and workers must be updated with the policy network. But at that point one shouldn't use the DataLoader APIs, but use the distributed APIs instead.

> develop a set of reinforcement learning algorithm classes which can be applied to those dataloaders, and populate policy values, based on rewards and such

This should definitely not be in core. We don't really provide algorithms, other than the ones implementing basic ops, optimizers. Algorithms should be built upon pytorch, not provided by the core library. Some high quality choices:
+ https://github.com/vitchyr/rlkit
+ https://github.com/DLR-RM/stable-baselines3

**Comment by alexge233** [comment]
Sorry it took so long to reply.

I agree, having looked at the links you've provided, I'll just keep this into a separate repository, will probably still end up using ATen, and would like to conform to how PyTorch does things, but it seems unnecessary to place it in PyTorch.
Feel free to close this.
